src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -169,20 +173,57 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 {
                     if (NetEventSource.IsEnabled) Trace($"Failed to send request body: {e}");
 
-                    // Cancel the stream before we set _requestCompletionState below.
-                    // Otherwise, a response stream reader may race with the actual Cancel.
-                    Cancel();
+                    bool signalWaiter = false;
 
                     lock (SyncObject)
                     {
                         Debug.Assert(_requestCompletionState == StreamCompletionState.InProgress, $"Request already completed with state={_requestCompletionState}");
 
-                        _requestCompletionState = StreamCompletionState.Failed;
+                        if (_requestBodyAbandoned)
+                        {
+                            // See comments on _requestBodyAbandoned.
+                            // In this case, the request is still considered successful and we do not want to send a RST_STREAM, 
+                            // and we also don't want to propagate any error to the caller, in particular for non-duplex scenarios.
+                            Debug.Assert(_responseCompletionState == StreamCompletionState.Completed);
+                            _requestCompletionState = StreamCompletionState.Completed;
+                            Complete();
+                            return;
+                        }
+                        else
```

Comment:
Nit: there's a lot of code in this else block now.  Consider removing the `else` and unindenting the rest, since the `if` has a `return`.